### PR TITLE
Authentication: Login to comment

### PR DESCRIPTION
**Pull request by chenzizhao**
To prevent spamming and associate comments with users, we add user authentication. Only after logging in with their Google account can the users submit a comment or delete comments from the database. 

A new `LoginServlet` is created based on Google's Users API from App Engine. A GET request returns the login status ("true"/"false") and a POST request returns a string of the link to log in/out based on the current status. The front end checks the login status when reloaded. 

Comments and likes are still anonymous.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/LikeServlet.java [hunk 1]
```diff
@@ -39,17 +39,12 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
-        commentEntity = datastore.get(likeKey);        
+        buildTransaction(datastore, likeKey, txn);
       } catch (EntityNotFoundException e) {
         response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Comment not found.");
         break;
       }
-      long count = (long) commentEntity.getProperty("likeCount");
-      ++count;
-      commentEntity.setProperty("likeCount", count);
-      datastore.put(txn, commentEntity);
```
Comment (juanalonsocruz) [inline portfolio/src/main/java/com/google/sps/servlets/LikeServlet.java:52]: Nit: This shouldn't be part of this PR.

File: portfolio/src/main/java/com/google/sps/servlets/LikeServlet.java [hunk 2]
```diff
@@ -65,4 +60,18 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+    try {
+      commentEntity = datastore.get(key);        
+    } catch (EntityNotFoundException e) {
+      throw e;
+    }
+    long count = (long) commentEntity.getProperty("likeCount");
+    ++count;
+    commentEntity.setProperty("likeCount", count);
+    datastore.put(txn, commentEntity);
+  }
```
Comment (juanalonsocruz) [inline portfolio/src/main/java/com/google/sps/servlets/LikeServlet.java:76]: Nit: This shouldn't be part of this PR.

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+      String urlToRedirectToAfterUserLogsOut = "/";
+      String logoutUrl = userService.createLogoutURL(urlToRedirectToAfterUserLogsOut);
+      response.getWriter().println(logoutUrl);
+    } else {
+      String urlToRedirectToAfterUserLogsIn = "/";
+      String loginUrl = userService.createLoginURL(urlToRedirectToAfterUserLogsIn);
+      response.getWriter().println(loginUrl);
+    }
+  }
+}
```
Comment (juanalonsocruz) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:36]: POST should be used to create new resources (like adding a new entity to your database). GET should be used to get data without modifying data in the backend.
https://www.restapitutorial.com/lessons/httpmethods.html

Maybe you could get your GET to both things you want. You could make it return a JSON object with a boolean for whether they are logged in or not and also an action url (that would be either login or logout based on the other boolean).

File: portfolio/src/main/webapp/index.html [hunk 4]
```diff
@@ -8,8 +8,11 @@
   <script src="script.js"></script>
 </head>
 
-<body onload="getComments()">
+<body onload="getComments();checkLoginStatus();">
```
Comment (juanalonsocruz) [inline portfolio/src/main/webapp/index.html:11]: onload should be clean and only have one thing. Maybe create a new method `loadPage` or something like that calls all the methods needed to load more stuff in the page. Think of it like a `main` method.

File: portfolio/src/main/webapp/index.html [hunk 5]
```diff
@@ -30,10 +33,8 @@ <h2>Help me <button onclick="grow()">Grow</button></h2>
     <p id="endMsg"></p>
     <br>
     <h2>Comments</h2>
-    <form>
-      <label>Leave a comment </label>
-      <input type="text" id="new-comment" name="new-comment">
-      <button type="button" onclick="submitComment()">Submit</button>
+    <form id="leave-a-comment">
```
Comment (juanalonsocruz) [inline portfolio/src/main/webapp/index.html:36]: id should be a noun and not a verb. Maybe something like `comment-submission-form`. Change variables related to this in JS as well.

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/login")
+public class LoginServlet extends HttpServlet {
+  private UserService userService = UserServiceFactory.getUserService();
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("text/html;");
```
Comment (juanalonsocruz) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:18]: In general we should be sending JSON unless you are sending explicit html to the client. I see your js fetch reads it as json (i wonder if some magic is happening there to make it work lol) so this probably a small oversight.

File: portfolio/src/main/webapp/script.js [hunk 7]
```diff
@@ -100,4 +100,67 @@ function submitComment() {
 function likeComment(commentId) {
   const request = new Request(`/like?commentId=${commentId}`, { method: 'POST' })
   fetch(request).then(() => getComments());
+}
+
+function checkLoginStatus() {
```
Comment (juanalonsocruz) [inline portfolio/src/main/webapp/script.js:105]: Name should be descriptive of what the method actually does. Maybe something like `addLoginStatus`.

File: portfolio/src/main/webapp/script.js [hunk 8]
```diff
@@ -100,4 +100,67 @@ function submitComment() {
...
+  const request = new Request('/login', { method: 'GET' });
+  fetch(request)
+    .then(response => response.json())
+    .then(isLoggedIn => {
+      const deleteButtonContainer = document.getElementById('delete-container');
+      const leaveCommentForm = document.getElementById('leave-a-comment');
+      const loginContainer = document.getElementById('login-container');
+      if (isLoggedIn) {
+        buildLoggedInPage(deleteButtonContainer, leaveCommentForm, loginContainer)
```
Comment (juanalonsocruz) [inline portfolio/src/main/webapp/script.js:114]: Its not just building the element it is also appending it. I also think its better to have method that creates the element and the caller appends to the parent element.
For the isLoggedIn case you can start with `deleteButtonContainer.append(createDeleteCommentsButton());` and then follow this pattern with the rest making necessary helper methods that create more complex groups of elements.

File: portfolio/src/main/java/com/google/sps/servlets/LikeServlet.java (hunk 1 above)
Comment (chenzizhao) [reply portfolio/src/main/java/com/google/sps/servlets/LikeServlet.java:52]: emmm looks like I pushed to the wrong PR 😅

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 3 above)
Comment (chenzizhao) [reply portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java]: I was being lazy (didn't want to create a new class) and somehow managed to convince myself that "login" is to write the user's status (??). Now fixed lol

File: portfolio/src/main/webapp/index.html (hunk 4 above)
Comment (chenzizhao) [reply portfolio/src/main/webapp/index.html]: Done.

Actually I thought about that and then "nah" -- maybe from now on I should re-calibrate. When I find something ugly but too lazy to change, I should refactor the code.

File: portfolio/src/main/webapp/index.html (hunk 5 above)
Comment (chenzizhao) [reply portfolio/src/main/webapp/index.html]: 👌 will stick to the style guide

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 6 above)
Comment (chenzizhao) [reply portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java]: Done. 

Seems like `response.json()` doesn't check content type. I was sending a single boolean value, so `Boolean.toString()` and `Gson.toJson()` probably gave the same thing.

File: portfolio/src/main/webapp/script.js (hunk 7 above)
Comment (chenzizhao) [reply portfolio/src/main/webapp/script.js]: 👌

File: portfolio/src/main/webapp/script.js (hunk 8 above)
Comment (chenzizhao) [reply portfolio/src/main/webapp/script.js]: Yap done! Big refactoring but it also allows me to reuse `createButton()` to create the thumbs up buttons.
